text/chapter10.md

```diff
@@ -503,36 +503,56 @@ exports.unsafeHead = arr => {
 
 ## Using Type Class Member Functions
 
-Just like our earlier guide on passing the `Maybe` constructor over FFI, this is another case of writing PureScript that calls JavaScript, which in turn calls PureScript functions again. Here we will explore how to pass type class member functions over the FFI.
+Say we want to write a function that relies on its argument having a `Show` instance, in a foreign JavaScript module. We might want to do this:
 
-We start with writing a foreign JavaScript function which expects the appropriate instance of `show` to match the type of `x`.
+```hs
+-- Don't do this.
+
+foreign import boldBad :: forall a. Show a => a -> String
+```
+
+This would work provided the foreign module exports:
 
 ```js
-exports.boldImpl = show => x =>
-  show(x).toUpperCase() + "!!!";
+// Seriously. Don't. Read on.
+
+exports.boldBad = Show => x =>
+  Show.show(x).toUppercase() + "!!!";
 ```
 
-Then we write the matching signature:
+A constraint in a foreign import definition results in importing a curried function that accepts a _dictionary_ object containing the class members. The compiler implicitly passes, for us, a dictionary of the correct instance members when calling the foreign function. However, this practice has been DEPRECATED as representing type class instances as dictionaries is an accidental complexity. Doing so will trigger a warning.
```

Comment:
```suggestion
A constraint in a foreign import definition results in importing a curried function that accepts a _dictionary_ object containing the class members. The compiler implicitly passes, for us, a dictionary of the correct instance members when calling the foreign function. However, this practice has been DEPRECATED and doing so will trigger a warning. Representing type class instances as dictionaries is an implementation detail that may change in the future. For example, the compiler may eventually elide type class dictionaries in some cases, and we wouldn't want this optimization to break existing code.
```

Reply:
I don't think the prospect (or lack) of optimization is the reason to give here. Although it was one of the driving motivations.

The fact that dictionaries are a _leaked implementation detail_ is just (likely correct) speculation. Had type classes weren't implemented as records, FFI could still have chosen to pass members as dictionaries.

Point is, use of dictionaries was probably just a weak API choice. There is just no reason to do so. Probably a convenience at the time. Maybe shorter import declarations were preferred.

We don't need to justify the deprecation here. Just need to state that it is. And a reason that it was.

What do you thing of the latest commit?

Reply:
> The fact that dictionaries are a leaked implementation detail is just (likely correct) speculation.

Don't know what you mean by _speculation_. This claim is backed-up by folks more familiar with the compiler's development history:
> We generally consider passing dictionaries as an implementation detail that users shouldn't rely on for their FFI as specialization optimizations would break FFI code written like that.

https://github.com/purescript/documentation/issues/273

I think describing what may cause FFI code to break in the future (e.g. optimizations) makes this rule more memorable for readers.

Reply:
> Don't know what you mean by _speculation_.

Exactly that. There is no evidence that it _is_. It very much looks like so. It is generally considered. It would be inaccurate to claim so. And even if it were (which I too believe is), it is not the _reason_ of deprecation. It's an explanation of the how things became.

This [comment](https://github.com/purescript/purescript/issues/3182#issuecomment-582725320) caught my attention.

> I just think it should be removed because I can't think of a place where it was a good idea,

There is no intrinsic reason of why type class members be provided via its _representation_ (which is a dictionary). Having done so proved a number of issues including
  - reliance on implementation detail, i.e. member names
  - hinder optimization

Type class representation in other parts of the language could change whiles still maintaining the dictionary API for FFI. For optimizing calling foreign functions, there would remain the case class with no members still require an empty dictionary. These are doable but probably more complex. Than if FFI were striped access to such dictionaries.

Taking a step back, explicitly passing needed functions (type class members or not) is the _simpler_ approach. At the FFI level it's just higher-ordered functions. There is no loss in giving up this feature.

Even if optimization was the practical reason for deprecation. (I too believe so) I think its beyond-scope to mention it. Optimization is an implementation detail worth hiding.

> I think describing what may cause FFI code to break in the future (e.g. optimizations) makes this rule more memorable for readers.

I think the goal of this section is just to note existence of this deprecated feature. Because it is likely that one will try. I did. When they do, they'll remember. If they never do, it doesn't matter that they don't.

Reply:
> There is no evidence that it _is_. It very much looks like so. It is generally considered. It would be inaccurate to claim so.

I don't see how describing type class dictionaries as an implementation detail is _inaccurate_. That's how they've been [described](https://github.com/purescript-contrib/purescript-book/pull/140#discussion_r425621621) to me. I still think a similar explanation would be helpful for readers (if we even want to cover this "what not to do" topic in the book).

@hdgarrood, do you have any suggestions on how to proceed?

Reply:
The line between what counts as an implementation detail and what doesn't is a little blurry for sure, but if we define "implementation details" as "something which isn't considered part of the public API and can therefore change without warning," then it's not actually _that_ blurry, especially in this case where we have explicit statements from maintainers - it does just depend on what the maintainers say, and what changes maintainers consider to be breaking. I think the fact that `Show a => a -> String` becomes a curried JS function of two arguments, where the first is a Show dictionary, is unlikely to change and is probably fine to rely on, but the _representation_ of dictionaries is certainly not. So if `showDict` is a Show dictionary, you shouldn't rely on being able to write `showDict.show(something)`.

I disagree that the deprecated approach that we are trying to warn users away from in this PR is obvious. I think many users, especially beginners, don't really understand how type class dictionaries work under the surface (which is fine; you don't usually need to, especially when you're still getting started). If you don't understand how type classes work under the surface, then it's hard to imagine how it would occur to you to rely on their runtime representation. I suspect most instances of this pattern in the wild arose because we previously had a page in the purescript/documentation repo recommending it and telling people how to do it.

For me, the fact that

* this pattern is not explicitly discussed in the book,
* the compiler produces a deprecation warning when you try to do it,
* the old page in the purescript/documentation repo has now been updated to discourage people from doing this

is sufficient. I don't think we additionally need a paragraph in the book to tell people not to do it.

Reply:
> I disagree that the deprecated approach that we are trying to warn users away from in this PR is obvious.

A beginner will obviously not succeed, but likely try. I recall doing this.
```purescript
-- this is the type I want for my function
 foreign import showIt :: forall a. Show a => a -> String
```
```js
exports.showIt(a) {
  // tried show(a);
  // tried a.show();
  // don't know what to do... just return arbitrary string then
  return "ok";
}
```
```purescript
main = log $ showIt 4 -- print "ok" ?
```
This blew up alright. But it compiled! What's going on !? This can turn into a rabbit hole real fast if not warned, in spite the presence of the deprecate warning. We're not exactly steering people away from bad practice/risk of breakage. Just information on a deprecated, but exists, feature.

So I'm still for including this to the book. I see a couple of ways to move forward.
  1. Just stop at the foreign import declaration. As it would be had this feature been removed already. You might want to do this ... , but don't. Constraints in foreign import declarations are deprecated and will soon be removed.
  2. Give a more substantial reason for the deprecation.

I'm inclined to do the latter. Though we can agree that the deprecated feature exposes the _representation_ of type class dictionaries &mdash; an implementation detail that should be private, we could have mitigated this exposure in others ways, say defining a specification of such dictionaries for use in FFI. The choice to, eventually, drop this feature &mdash; to my understanding &mdash; are based on
  1. Passing dictionaries to foreign functions weren't necessary in the first place.
  2. It eliminates an optimization problem &mdash; no dictionaries, no empty class case.

So how about:
> Accessing type class members though dictionary objects was an unnecessary indirection. Not passing dictionaries also enhances code optimization.

Reply:
I’m still not convinced, personally. If we included every single thing a beginner might be tempted to do in order to warn against it, the book would balloon in size very quickly, and any extra discussion of this issue is of no value to someone who wouldn’t try it in the first place.

Reply:
@hdgarrood I agree we can't account for every possible mistake. And that this is valuable only to those who might try. Are you implying that this particular mistake is not worth warning about. That attempting to type an imported function with constraints the same way we put constraints on purescript functions is not the first thing that comes to mind. And it will go unnoticed had we just went ahead with the proper way.

Would these guidelines also apply to [this section](https://book.purescript.org/chapter10.html#beyond-simple-types)?

> ```javascript
> // Don't do this
> 
> let Data_Maybe = require("../Data.Maybe")
> 
> exports.maybeHead = arr => {
>   if (arr.length) {
>     return Data_Maybe.Just.create(arr[0]);
>   } else {
>     return Data_Maybe.Nothing.value;
>   }
> }
> ```

Reply:
That's a good question. I'll have a think.